model/map/src/main/java/org/keycloak/models/map/storage/tree/DefaultTreeNode.java

```diff
@@ -303,4 +342,73 @@ public void setParent(Self parent) {
     private Self getThis() {
         return (Self) this;
     }
+
+    @Override
+    public int hashCode() {
+        return this.uniqueId;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this == obj;
+    }
+
+    @Override
+    public Stream<Self> getParentsStream() {
+        Builder<Self> resBuilder = Stream.builder();
+        for (Optional<Self> p = getParent(); p.isPresent(); p = p.get().getParent()) {
+            resBuilder.accept(p.get());
+        }
+        return resBuilder.build();
+    }
+
+    private static final ThreadLocal<Boolean> TOSTRING_DETAILS = new ThreadLocal<Boolean>() {
+        @Override
+        protected Boolean initialValue() {
+            return Boolean.TRUE;
+        }
+
+    };
+
+    /**
+     * Print a tree structure in a pretty ASCII format.
+     * Adopted from https://stackoverflow.com/a/53705889/6930869
+     *
+     * @param prefix Current prefix. Use "" in initial call
+     * @param node The current node
+     * @param getChildrenFunc A {@link Function} that returns the children of a given node.
+     * @param isTail Is node the last of its siblings. Use true in initial call. (This is needed for pretty printing.)
+     * @param <T> The type of your nodes. Anything that has a toString can be used.
+     */
+    private StringBuilder toString(StringBuilder output, String prefix, Self node, boolean isTail) {
+        String nodeName = node.getLabel();
+        if (Objects.equals(TOSTRING_DETAILS.get(), Boolean.FALSE)) {
+            return new StringBuilder("@").append(nodeName);
+        }
+        String nodeConnection = isTail ? (prefix.isEmpty() ? "O── " : "└── ") : "├── ";
+        output.append(prefix).append(nodeConnection).append(nodeName);
+        try {
+            TOSTRING_DETAILS.set(Boolean.FALSE);
+            output.append(nodeProperties.isEmpty() ? "" : " " + nodeProperties);
```

Comment:
```suggestion
            output.append(node.getNodeProperties().isEmpty() ? "" : " " + node.getNodeProperties());
            
```
Maybe this is what Michal suggests?